content/en/docs/plugins/developers_guide.md

```diff
@@ -0,0 +1,1081 @@
+
+`Init()` method is called to initialize a plugin when the framework allocates it. A user-defined configuration string is passed by the framework. This is were the plugin can initialize its internal state and acquire all the resource it needs.
+
+`Destroy()` is an optional method that can be used to release the resources used by a plugin before it gets destroyed by the framework.
+
+`Fields()` returns an array of `sdk.FieldEntry` representing all the fields supported by a plugin for extraction. The order of the fields is relevant, as their index is used as an identifier during extraction.
+
+`Extract()` extracts the value of one of the supported fields from a given event passed in by the framework. The `sdk.ExtractRequest` argument should be used to set the extracted value.
+
+#### Defining Source Plugins
+In the Go SDK, a source Plugin must specify two types, one of the plugin itself and one for the plugin instances, implementing the following interfaces respectively:
+```go
+// sdk/plugins/source
+type Plugin interface {
+	...
+	Info() *sdk.Info
+	Init(config string) error
+	// Destroy()
+	Open(params string) (Instance, error)
+	String(in io.ReadSeeker) (string, error)
+	// Fields() []sdk.FieldEntry
+	// Extract(req sdk.ExtractRequest, evt sdk.EventReader) error
+}
+
+// sdk/plugins/source
+type Instance interface {
+	...
+	// Close()
+	// Progress(pState PluginState) (float64, string)
+	NextBatch(pState PluginState, evts EventWriters) (int, error)
+}
+```
+The `source.Plugin` interface has many functions in common with `extractor.Plugin`. One difference is that the extraction-related functions `Fields()` and `Extract()` are optional, and must be defined only if the source plugin enables the optional extraction capabilities. Instead, the two `Open()` and `String()` methods are mandatory.
+
+`Open()` creates a new plugin instance to open a new stream of events. The framework provides the user-defined open parameters to customize the event source. The return value must implement the `source.Instance` interface, and its lifecycle ends when the event stream is closed.
+
+`String()` returns a string representation of an event created by the plugin. The string representation should be on a single line and contain important information about the event. 
+
+The `source.Instance` interface represents plugin instances for an opened event stream, and has one mandatory method and few optional ones.
+
+`NextBatch` creates a new batch of events to be pushed in the event stream. The SDK provides a pre-allocated batch to write events into, in order to manage the used memory optimally.
+
+`Close()` is an optional method that can be used to release the resources used by a plugin instance before its event stream gets closed.
+
+`Progress()` is an optional method returning a float and string representation of the current progress of the event stream. This applies to all the event stream for which the total number of events can be determined.
```

Comment:
I would mention the units here. Is it 0-100% and the values are converted back to 0-10000 as the underlying API uses, or should the values be 0-10000 here (in which case they should probably be an int, not float).

Reply:
thanks, looks better!